### PR TITLE
feat(ora|h3): add H3 spatial indexing module for Oracle

### PR DESCRIPTION
**Pull request by diego-vicente**
# Description

Shortcut

- Story: https://app.shortcut.com/cartoteam/story/548739
- Autolink: [sc-548739]

This PR adds the H3 spatial indexing module to Oracle. It ports all 19 public functions from the existing BigQuery/Snowflake/Redshift implementations.

All functions are implemented as PL/SQL stored functions, leveraging Oracle's native `SDO_UTIL.H3_*` built-ins for geometry operations (boundary, center, validity checks) and implementing traversal/hierarchy logic (kring, hexring, compact, polyfill) in pure PL/SQL. Collections are passed as JSON arrays since Oracle does not support array-typed function parameters.

**Note**: `oracle_db.py` was updated to cache the extracted wallet directory at module level, preventing TNS name invalidation when `oracledb` reuses its internal connection cache across multiple test connections. Previously, each call to `extract_wallet()` created a new temp directory that could be cleaned up while connections still referenced it.

## Type of change

- Feature
- Tests

# Acceptance

138 integration tests covering all 19 functions, including edge cases for pentagon cells, resolution boundaries (0 and 15), cross-meridian geometries, compact/uncompact round-trips, polyfill with MultiPolygon/holes, and null handling.

# Basic checklist

- [X] Good PR name
- [ ] Shortcut link
- [X] Just one issue per PR
- [ ] GitHub labels
- [ ] Proper status & reviewers
- [X] Tests
- [ ] Documentation

### CODE REVIEW

File: clouds/oracle/common/oracle_db.py [hunk 1]
```diff
@@ -10,12 +10,34 @@
...
+
+atexit.register(_cleanup_wallet)
+
 
 def extract_wallet(wallet_zip_b64, wallet_password):
     """Extract Oracle wallet from base64-encoded ZIP."""
+    global _wallet_dir, _conn_string
+
+    if _wallet_dir and os.path.exists(_wallet_dir):
+        return _wallet_dir, _conn_string
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/common/oracle_db.py:38]: This is test infrastructure running under single-threaded pytest — there's no concurrent access to `extract_wallet`. Adding a threading lock here would be YAGNI.

File: clouds/oracle/modules/sql/h3/H3_FROMGEOGPOINT.sql [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    END IF;
+    IF resolution < MIN_RESOLUTION OR resolution > MAX_RESOLUTION THEN
+        RETURN NULL;
+    END IF;
+    IF point.GET_GTYPE() != POINT_GTYPE THEN
+        RETURN NULL;
+    END IF;
+
+    h3_raw := SDO_UTIL.H3_KEY(point, resolution);
+    RETURN LOWER(LTRIM(RAWTOHEX(h3_raw), '0'));
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_FROMGEOGPOINT.sql:28]: H3 index 0 is not a valid H3 cell. Valid H3 indices always have mode bits set in the high bytes (mode 1 for cells), so `RAWTOHEX` of any valid cell will never produce all zeros. `SDO_UTIL.H3_KEY` only returns valid cells, so this edge case cannot occur in practice.

File: clouds/oracle/modules/sql/h3/H3_HEXRING.sql [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                RAISE;
+            END IF;
+            RAISE_APPLICATION_ERROR(ERR_INVALID_ORIGIN, 'Invalid input origin');
+    END;
+
+    cell_res := SDO_UTIL.H3_RESOLUTION(h3_raw);
+
+    -- Special case: distance 0 returns just the origin
+    IF dist = 0 THEN
+        RETURN '["' || origin || '"]';
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_HEXRING.sql:93]: The H3 library's `gridRing(origin, 0)` returns `[origin]`, not `[]`. All other cloud implementations in this repo (BigQuery, Snowflake, Postgres) return `[origin]` for distance=0, and our tests verify this behavior. The current code is correct.

File: clouds/oracle/modules/sql/h3/H3_POLYFILL_TABLE.sql [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    -- Extract schema from output_table for function reference
+    IF INSTR(output_table, '.') > 0 THEN
+        v_schema := SUBSTR(output_table, 1, INSTR(output_table, '.') - 1);
+    ELSE
+        v_schema := SYS_CONTEXT('USERENV', 'CURRENT_SCHEMA');
+    END IF;
+
+    -- Build CTAS using JSON_TABLE to unnest the polyfill results
+    -- The input_query must have a column named 'geom' of type SDO_GEOMETRY
+    v_sql := 'CREATE TABLE ' || output_table || ' AS '
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_POLYFILL_TABLE.sql]: Fixed — added `DBMS_ASSERT.QUALIFIED_SQL_NAME(output_table)` to sanitize the table name in the dynamic SQL statement.

File: clouds/oracle/modules/sql/h3/H3_HEXRING.sql [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    -- JSON building
+    json_result VARCHAR2(32767);
+    first_entry BOOLEAN;
+    v_key VARCHAR2(16);
+BEGIN
+    -- Validate inputs
+    IF origin IS NULL THEN
+        RAISE_APPLICATION_ERROR(ERR_INVALID_ORIGIN, 'Invalid input origin');
+    END IF;
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_HEXRING.sql:63]: All cloud implementations in this repo raise errors for NULL origin in HEXRING/KRING. Switching to return `[]` would be a cross-cloud consistency divergence.

File: clouds/oracle/modules/sql/h3/H3_POLYFILL.sql [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        END IF;
+        json_result := json_result || '"' || v_key || '"';
+        v_key := result_map.NEXT(v_key);
+    END LOOP;
+    json_result := json_result || ']';
+
+    -- Return NULL instead of empty array for consistency with other clouds
+    IF result_map.COUNT = 0 THEN
+        RETURN NULL;
+    END IF;
```
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_POLYFILL.sql:229]: This was a deliberate choice — the code comment says "Return NULL instead of empty array for consistency with other clouds." BigQuery and Postgres also return NULL for empty polyfill results. Snowflake is the outlier returning `[]`.

**Comment by vdelacruzb** [comment]
## PR Review: Typing + pending items

> **Updated** to align with the Oracle typing convention ratified in #602. Prior version of this comment is superseded by the sections below.

### 1. Return type for array-returning functions

Seven functions currently return `VARCHAR2`/`CLOB` JSON: `H3_TOCHILDREN`, `H3_KRING`, `H3_HEXRING`, `H3_KRING_DISTANCES`, `H3_UNCOMPACT`, `H3_COMPACT`, `H3_POLYFILL`.

Per the convention in #602, these refactor to nested table types + pipelined functions:

```sql
-- clouds/oracle/modules/sql/h3/00_types.sql (new file)
CREATE OR REPLACE TYPE @@ORA_SCHEMA@@.H3_INDEX_ARRAY
    AS TABLE OF VARCHAR2(16);
CREATE OR REPLACE TYPE @@ORA_SCHEMA@@.H3_DISTANCE_PAIR AS OBJECT (
    h3_index VARCHAR2(16), distance NUMBER
);
CREATE OR REPLACE TYPE @@ORA_SCHEMA@@.H3_DISTANCE_ARRAY
    AS TABLE OF @@ORA_SCHEMA@@.H3_DISTANCE_PAIR;

-- Function
CREATE OR REPLACE FUNCTION @@ORA_SCHEMA@@.H3_TOCHILDREN(
    h3 VARCHAR2, res NUMBER
) RETURN @@ORA_SCHEMA@@.H3_INDEX_ARRAY PIPELINED AS
BEGIN
    ...
    PIPE ROW(child_hex);
    RETURN;
END;

-- Consumer — mirrors BQ UNNEST / SF FLATTEN / PG UNNEST
SELECT column_value AS h3
FROM TABLE(CARTO.H3_TOCHILDREN('85283473fffffff', 10));
```

### 2. Logic verification — what works ✓

- **Geometry SRID**: all user-visible geometry outputs are 4326. `H3_BOUNDARY`/`H3_CENTER` delegate to `SDO_UTIL.H3_*` (4326). `H3_FROMLONGLAT`, `H3_KRING`, `H3_HEXRING`, `H3_KRING_DISTANCES`, `H3_DISTANCE` construct points with hardcoded `SRID_WGS84`. The `geom_srid` preservation inside `H3_POLYFILL`/`__H3_POLYFILL_MODE` is internal only (to avoid `SDO_GEOM.RELATE` mismatch) — no geometry escapes to the caller.
- **Algorithm correctness**: `H3_TOCHILDREN` bit-manipulation, `H3_POLYFILL` grid sampling, pentagon handling all produce the same cell IDs as BigQuery.
- **Test depth** actually exceeds BQ/SF (13 tests for `H3_TOCHILDREN` vs 2, 11 for `H3_POLYFILL` vs 2, explicit null/boundary cases).

### 3. Pending items

**🔴 Blockers before merge (aligned with #602):**

| # | Item | Effort |
|---|---|---|
| 1 | Create `clouds/oracle/modules/sql/h3/00_types.sql` with `H3_INDEX_ARRAY`, `H3_DISTANCE_PAIR`, `H3_DISTANCE_ARRAY` | 1 h |
| 2 | Refactor 8 array-returning functions (7 public + `__H3_POLYFILL_MODE`) to `H3_INDEX_ARRAY` / `H3_DISTANCE_ARRAY` PIPELINED | 1 day |
| 3 | Refactor `H3_POLYFILL_TABLE` to consume `TABLE(...)` internally instead of `JSON_TABLE(...)` | 1–2 h |
| 4 | Migrate tests to `SELECT COLUMN_VALUE FROM TABLE(func(...))` pattern; drop `json.loads(...)` | 2–3 h |
| 5 | Add `oracle` to `VALID_CLOUDS` in `core/Makefile` | 5 min |
| 6 | Create 20 `.md` docs in `clouds/oracle/modules/doc/h3/` + `_INTRO.md` with `TABLE(...)` consumer examples | 3–4 h |

**🟡 Should-fix:**

- Reuse `test_utils.drop_table` (added in #601) in `test_H3_POLYFILL_TABLE.py` — currently has a private `_drop_table` helper at 8 call sites.
- `H3_FROMGEOGPOINT` silently accepts non-WGS84 input SRID — decide policy (reject, auto-transform, or document requirement) and test.
- Add SRID=4326 assertion tests for `H3_FROMLONGLAT`, `H3_KRING`, `H3_HEXRING`, `H3_KRING_DISTANCES` (pattern already exists in `test_H3_BOUNDARY.py:44-52`).
- Add direct tests for `__H3_POLYFILL_MODE` with `mode='intersects'` and `mode='center'` (currently only exercised indirectly via `H3_POLYFILL_TABLE`).
- Change `H3_ISVALID` / `H3_ISPENTAGON` return type from bare `NUMBER` to `NUMBER(1)` per the updated rule in #602 (single-digit constraint at the schema level). Document the `NUMBER(1)` vs `BOOL` divergence in the function docs — callers must use `= 1`, not truthy predicates.

**🟢 Follow-up (non-blocking, can land after v1.0):**

- Backport Oracle's deeper test patterns (roundtrip, explicit null/boundary cases) to BQ/SF.
- Add missing `H3_UNCOMPACT` tests in BQ/SF (function exists, no test file).
- Add missing `H3_POLYFILL_TABLE.test.js` in Snowflake.
- Add 3-arg `H3_POLYFILL(geom, res, mode)` public overload (currently mode is only accessible through `H3_POLYFILL_TABLE`).

**Dependency**: #602 should merge first (or fold into this PR) so the rules file is canonical before this module ships.

**Comment by vdelacruzb** [comment]
## Admin — missing shortcut link and labels

A few process items from the PR's own "Basic checklist" still pending:

- 🔴 **Shortcut link missing** — description still contains the template placeholder: `Story: https://app.shortcut.com/cartoteam/story/STORY_ID` and `Autolink: [sc-STORY_ID]`. Needs the real story ID and autolink so the PR shows up against the epic.
- 🟡 **No GitHub labels** assigned.
- 🟡 **No milestone / project / epic** attached.
- 🟡 **"Proper status & reviewers"** unchecked in the PR's own checklist.

None of these block code review, but they matter for release notes, epic tracking, and automated Shortcut ↔ GitHub linking. Fixable inline in the PR description + GitHub sidebar without a commit.

**Comment by diego-vicente** [comment]
## Update — addressed in latest commits (e5a615d1..bd097656)

All 🔴 blockers and 🟡 should-fix items from this review have landed:

### 🔴 Blockers — done

| # | Item | Commit |
|---|---|---|
| 1 | `00_H3_INDEX_ARRAY.sql` + `00_H3_DISTANCE_ARRAY.sql` (with `H3_DISTANCE_PAIR` object) | e5a615d1 |
| 2 | 7 array-returning functions + `__H3_POLYFILL_MODE` refactored to `PIPELINED` returning `H3_INDEX_ARRAY` / `H3_DISTANCE_ARRAY` | 21dcfb2a |
| 3 | `H3_POLYFILL_TABLE` now consumes `TABLE(__H3_POLYFILL_MODE(…))` directly (no `JSON_TABLE`); resolution and mode bound via `USING`; identifier still sanitized with `DBMS_ASSERT.QUALIFIED_SQL_NAME` | 21dcfb2a |
| 4 | All h3 tests use `SELECT COLUMN_VALUE FROM TABLE(…)`; `json.loads` removed | 92147f3a |
| 5 | `oracle` added to `VALID_CLOUDS` in root `Makefile` | bd097656 |
| 6 | 20 docs in `clouds/oracle/modules/doc/h3/` (`_INTRO.md` + 19 function/procedure pages, including `H3_POLYFILL_TABLE`), cloned from the Postgres docs and adapted for Oracle types and `TABLE(…)` consumer syntax | c8e37a76 |

Type-file layout follows the **per-type `00_<TYPE>.sql`** rule from the merged convention rather than a single `00_types.sql`, with reverse-dependency `DROP … FORCE` for the collection-of-object case so redeploys don't trip ORA-02303.

`H3_COMPACT` / `H3_UNCOMPACT` accept `H3_INDEX_ARRAY` as input; `H3_UNCOMPACT` consumes `H3_TOCHILDREN` via `TABLE(…)` instead of round-tripping through `JSON_ARRAY_T`.

### 🟡 Should-fix — done

- `test_H3_POLYFILL_TABLE.py` uses `test_utils.drop_table` from #601 (8 call sites; the local helper is gone). 92147f3a
- `H3_FROMGEOGPOINT`: rejects geometries with an explicit non-4326 SRID (returns NULL); accepts NULL SRID and treats it as WGS84 (matches `SDO_UTIL.FROM_WKTGEOMETRY` output). Documented in the SQL header and in `H3_FROMGEOGPOINT.md`. 1c133ce6
- `__H3_POLYFILL_MODE` direct tests added in `test_H3_POLYFILL.py` for `center`, `intersects`, and an unknown-mode case. 92147f3a
- `H3_ISVALID` / `H3_ISPENTAGON` return `NUMBER(1)`; the docs note the `= 1` consumer pattern explicitly. 1c133ce6
- **NULL-on-invalid alignment with #602:** `H3_KRING` / `H3_HEXRING` / `H3_KRING_DISTANCES` switched from `RAISE_APPLICATION_ERROR` to returning no rows on invalid input. Tests updated.

### Skipped, with reason

- **SRID=4326 assertion tests for `H3_FROMLONGLAT` / `H3_KRING` / `H3_HEXRING` / `H3_KRING_DISTANCES`** — these functions return `VARCHAR2` (cell IDs), not `SDO_GEOMETRY`, so the `t.geom.SDO_SRID = 4326` pattern from `test_H3_BOUNDARY.py` doesn't apply. `H3_BOUNDARY` and `H3_CENTER` (the geometry-returning functions) already have those assertions in place.

File: clouds/oracle/common/oracle_db.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/common/oracle_db.py:38]: No change in the latest push — sticking with the YAGNI position. The wallet cache only runs under single-threaded pytest; we'll add a `threading.Lock` if/when we ever invoke it concurrently (e.g. from a worker pool).

File: clouds/oracle/modules/sql/h3/H3_FROMGEOGPOINT.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_FROMGEOGPOINT.sql:48]: No change to the `LTRIM('0')` pattern — `SDO_UTIL.H3_KEY` only returns valid H3 cells, which always have mode bits set in the high bytes, so `RAWTOHEX` of a valid cell is never all zeros. The H3 index 0 is unreachable here.

In 1c133ce6 this file did get a related fix on a separate concern: `H3_FROMGEOGPOINT` now rejects geometries with an explicit non-WGS84 SRID (returns NULL) instead of silently passing them through, per the SRID rule in #602.

File: clouds/oracle/modules/sql/h3/H3_HEXRING.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_HEXRING.sql]: No change for `distance = 0` — the BigQuery, Snowflake and Postgres implementations in this repo all return `[origin]` for k=0, and our tests verify it.

In 21dcfb2a the function did become pipelined: `H3_HEXRING` now returns `H3_INDEX_ARRAY PIPELINED` and the `distance = 0` branch becomes `PIPE ROW(LOWER(origin)); RETURN;` (one row, the origin). Behaviour is identical to before for the consumer.

File: clouds/oracle/modules/sql/h3/H3_POLYFILL_TABLE.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_POLYFILL_TABLE.sql]: Doubly mitigated as of 21dcfb2a. `DBMS_ASSERT.QUALIFIED_SQL_NAME` is still in place for the table identifier (DDL can't bind names), and the procedure now also routes `mode` and `resolution` through `EXECUTE IMMEDIATE … USING` bind variables instead of concatenating them — so the only thing concatenated into the dynamic SQL is the asserted identifier and the user-supplied `input_query` (which is the documented contract).

File: clouds/oracle/modules/sql/h3/H3_HEXRING.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_HEXRING.sql]: **Reversed in 21dcfb2a / 92147f3a.** Per the NULL-on-invalid rule codified in #602, `H3_KRING`, `H3_HEXRING` and `H3_KRING_DISTANCES` now return zero rows on invalid input (NULL origin, invalid origin, NULL/negative size) instead of raising `ORA-200xx`. Tests were updated to expect empty results rather than an exception.

File: clouds/oracle/modules/sql/h3/H3_POLYFILL.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (diego-vicente) [reply clouds/oracle/modules/sql/h3/H3_POLYFILL.sql]: Resolved structurally in 21dcfb2a. `H3_POLYFILL` is now pipelined (`RETURN H3_INDEX_ARRAY PIPELINED`) — there is no longer a NULL vs `'[]'` choice to make: an empty result simply pipes zero rows. Consumers always do `SELECT COLUMN_VALUE FROM TABLE(carto.H3_POLYFILL(...))` and treat empty as zero rows uniformly.

**Review by vdelacruzb — APPROVED** [review]
LGTM. I've adapted a bit the module by using the h3 library for the most complex functions.

Some small tests for the polyfill with `contains`and `intersects`:
Screenshot 2026-05-04 at 17 41 33
Screenshot 2026-05-04 at 17 42 58
